R/shiny.R

```diff
@@ -502,12 +653,157 @@ ShinySession <- R6Class(
         ns = ns,
         makeScope = function(namespace) {
           self$makeScope(ns(namespace))
+        },
+
+        setBookmarkExclude = function(names) {
+          bookmarkExclude <<- names
+        },
+        getBookmarkExclude = function() {
+          bookmarkExclude
+        },
+        onBookmark = function(fun) {
```

Comment:
They'll get invoked in the order that they were registered, so if the module has an `onBookmark` callback and`callModule()` is called before the app's own `onBookmark`, then the module's callback will be invoked first. But if the app calls `onBookmark()` and then `callModule()`, the app's callback will be invoked first. But the `state$values` code is written so that order won't matter.